src/main.rs

```diff
@@ -24,138 +38,117 @@ use std::path::PathBuf;
+            s.push(' ');
         }
     }
+    s
 }
 
-async fn main_loop(mut wm: WindowManager, mut stream: EventStream) {
-    loop {
-        if process_events(&mut wm, &mut stream).await.is_err() {
-            log::info!("Couldn't process WM events. The WM might have been terminated");
-            log::info!("Attempting to reconnect to the WM");
-            if let Ok((w, s)) = WindowManager::connect().await {
-                wm = w;
-                stream = s;
-                log::info!("Successfully reconnected to WM");
+fn process_events(config: &Config, mut wm: WindowManager, stream: EventStream) -> Result<()> {
+    for _event in stream {
+        let workspaces = wm.get_windows_in_each_workspace()?;
+        for (name, windows) in workspaces {
+            let new_name = pretty_windows(config, &windows);
+            let num = name
+                .split(':')
+                .next()
+                .ok_or(anyhow!("Unexpected workspace name"))?;
+            if new_name.is_empty() {
+                wm.rename_workspace(&name, num)?;
             } else {
-                log::info!("Failed to reconnect to WM. Will try again in 1 second");
-                std::thread::sleep(std::time::Duration::from_millis(1000));
+                wm.rename_workspace(&name, &format!("{num}: {new_name}"))?;
             }
         }
     }
+    bail!("Can't get next event")
 }
 
-#[tokio::main]
-async fn main() -> Result<(), &'static str> {
-    pretty_env_logger::init();
-    let _ = Args::parse();
-
+fn lockfile_path() -> PathBuf {
     let mut lockfile_path = match dirs::runtime_dir() {
         Some(path) => path,
         None => PathBuf::from("/tmp"),
     };
     lockfile_path.push("workstyle.lock");
+    lockfile_path
+}
 
-    let path_str = String::from(lockfile_path.to_str().unwrap());
+fn aquire_lock() {
+    // Try to aquire the lock
+    if LOCK.lock().unwrap().is_none() {
+        panic!("Failed to aquire the lock");
+    }
 
-    let lock = match Lockfile::create(lockfile_path) {
-        Ok(lock) => lock,
-        Err(err) => panic!("Unrecoverable error: {}, {}", err.into_inner(), path_str),
-    };
+    // Drop the lock on exit
+    let mut signals = Signals::new(TERM_SIGNALS).expect("Failed to create signals iterator");
+    spawn(move || {
+        let _ = signals.forever().next();
+        drop(LOCK.lock().unwrap().take());
+        exit(0);
+    });
+
+    // Drop the lock on panic
+    std::panic::set_hook(Box::new(|info| {
+        error!("{info}");
+        if let Ok(mut lock) = LOCK.lock() {
+            drop(lock.take());
+        }
+    }));
+}
+
+fn main() -> Result<()> {
+    let _ = Args::parse();
+    aquire_lock();
 
-    let signals = Signals::new(TERM_SIGNALS).expect("Failed to create Signals");
-    let handle = signals.handle();
-    let termination_signals_task = tokio::spawn(handle_signals(signals, lock));
+    env_logger::init();
+    let config = Config::new()?;
```

Comment:
By moving the config parsing outside of the loop, we are now forcing the user to reload workstyle on change to the config file.

In my opinion, it makes the user experience worse.

As a user, I like apps like alacritty that will autoreload their config file and I would prefer to keep that behaviour for workstyle.

Because parsing the config file is practically instant, I don't think the latency cost is noticeable from a human user's perspective. We can also be smarter and spawn its own thread that will parse it again on change to the config file if we don't want to reload it on any WindowManager event.

Reply:
Okay